File: restore_ip_address/mk/RestoreIpAddress.java
package leetcode;

import java.util.ArrayList;
import java.util.List;
import java.util.StringJoiner;

/**
 * @author manki.kim
 **/
public class RestoreIpAddress {
	public List<String> restoreIpAddresses(String s) {
		List<String> results = new ArrayList<>();
		final int START = 1;
		final int END = 3;
		for (int a = START; a <= END; a++)
			for (int b = START; b <= END; b++)
				for (int c = START; c <= END; c++)
					for (int d = START; d <= END; d++) {
						if (a + b + c + d == s.length()) {
							String first = s.substring(0, a);
							String second = s.substring(a, a + b);
							String third = s.substring(a + b, a + b + c);
							String fourth = s.substring(a + b + c, a + b + c + d);
							if (Integer.parseInt(first) > 255 || Integer.parseInt(second) > 255
								|| Integer.parseInt(third) > 255 || Integer.parseInt(fourth) > 255) {
								continue;
							}

							if (isValid(first) || isValid(second) || isValid(third) || isValid(fourth)) {
								continue;
							}
							StringJoiner joiner = new StringJoiner(".");
							joiner.add(first).add(second).add(third).add(fourth);

							String ip = joiner.toString();
							if (ip.length() == s.length() + 3) {
								results.add(ip);
							}
						}
					}

		return results;
	}

	private boolean isValid(String s) {
		return !(s.length() > 1 && s.startsWith("0"));
	}
}

File: restore_ip_address/thkim/RestoreIPAddress.kt
fun main(args:Array<String>){
    println(Solution().restoreIpAddresses("010010"))
}
class Solution {
    fun restoreIpAddresses(s: String): List<String> {
        val ipList = ArrayList<String>()
        for(first in 1 until 4){
            for(second in 1 until 4){
                for(third in 1 until 4){
                    for(fourth in 1 until 4){
                        if(first+second+third+fourth != s.length)
                            continue
                        try {
                            val octet1 = s.substring(0, first)
                            val octet2 = s.substring(first, first + second)
                            val octet3 = s.substring(first + second, first + second + third)
                            val octet4 = s.substring(first + second + third, first + second + third + fourth)
                            arrayListOf(octet1, octet2, octet3, octet4).let {
                                if(it.any { x-> x.toInt() > 255 || x.length >=2 && x.first() == '0'})
                                    throw Exception("Failed")
                                ipList.add("$octet1.$octet2.$octet3.$octet4")
                            }
                        }catch (e:Exception){
                            println(e.toString())
                        }
                    }
                }
            }
        }
        return ipList
    }
}